File: src/main/java/com/mssoftech/dbflute/bsentity/dbmeta/ProductDbm.java
package com.mssoftech.dbflute.bsentity.dbmeta;

import java.util.List;
import java.util.Map;

import org.dbflute.Entity;
import org.dbflute.optional.OptionalEntity;
import org.dbflute.dbmeta.AbstractDBMeta;
import org.dbflute.dbmeta.info.*;
import org.dbflute.dbmeta.name.*;
import org.dbflute.dbmeta.property.PropertyGateway;
import org.dbflute.dbway.DBDef;
import com.mssoftech.dbflute.allcommon.*;
import com.mssoftech.dbflute.exentity.*;

/**
 * The DB meta of product. (Singleton)
 * @author DBFlute(AutoGenerator)
 */
public class ProductDbm extends AbstractDBMeta {

    // ===================================================================================
    //                                                                           Singleton
    //                                                                           =========
    private static final ProductDbm _instance = new ProductDbm();
    private ProductDbm() {}
    public static ProductDbm getInstance() { return _instance; }

    // ===================================================================================
    //                                                                       Current DBDef
    //                                                                       =============
    public String getProjectName() { return DBCurrent.getInstance().projectName(); }
    public String getProjectPrefix() { return DBCurrent.getInstance().projectPrefix(); }
    public String getGenerationGapBasePrefix() { return DBCurrent.getInstance().generationGapBasePrefix(); }
    public DBDef getCurrentDBDef() { return DBCurrent.getInstance().currentDBDef(); }

    // ===================================================================================
    //                                                                    Property Gateway
    //                                                                    ================
    // -----------------------------------------------------
    //                                       Column Property
    //                                       ---------------
    protected final Map<String, PropertyGateway> _epgMap = newHashMap();
    { xsetupEpg(); }
    protected void xsetupEpg() {
        setupEpg(_epgMap, et -> ((Product)et).getProductId(), (et, vl) -> ((Product)et).setProductId(cti(vl)), "productId");
        setupEpg(_epgMap, et -> ((Product)et).getProductName(), (et, vl) -> ((Product)et).setProductName((String)vl), "productName");
        setupEpg(_epgMap, et -> ((Product)et).getProductHandleCode(), (et, vl) -> ((Product)et).setProductHandleCode((String)vl), "productHandleCode");
        setupEpg(_epgMap, et -> ((Product)et).getProductCategoryCode(), (et, vl) -> ((Product)et).setProductCategoryCode((String)vl), "productCategoryCode");
        setupEpg(_epgMap, et -> ((Product)et).getProductStatusCode(), (et, vl) -> ((Product)et).setProductStatusCode((String)vl), "productStatusCode");
        setupEpg(_epgMap, et -> ((Product)et).getRegularPrice(), (et, vl) -> ((Product)et).setRegularPrice(cti(vl)), "regularPrice");
        setupEpg(_epgMap, et -> ((Product)et).getRegisterDatetime(), (et, vl) -> ((Product)et).setRegisterDatetime(ctldt(vl)), "registerDatetime");
        setupEpg(_epgMap, et -> ((Product)et).getRegisterUser(), (et, vl) -> ((Product)et).setRegisterUser((String)vl), "registerUser");
        setupEpg(_epgMap, et -> ((Product)et).getRegisterProcess(), (et, vl) -> ((Product)et).setRegisterProcess((String)vl), "registerProcess");
        setupEpg(_epgMap, et -> ((Product)et).getUpdateDatetime(), (et, vl) -> ((Product)et).setUpdateDatetime(ctldt(vl)), "updateDatetime");
        setupEpg(_epgMap, et -> ((Product)et).getUpdateUser(), (et, vl) -> ((Product)et).setUpdateUser((String)vl), "updateUser");
        setupEpg(_epgMap, et -> ((Product)et).getUpdateProcess(), (et, vl) -> ((Product)et).setUpdateProcess((String)vl), "updateProcess");
        setupEpg(_epgMap, et -> ((Product)et).getVersionNo(), (et, vl) -> ((Product)et).setVersionNo(ctl(vl)), "versionNo");
    }
    public PropertyGateway findPropertyGateway(String prop)
    { return doFindEpg(_epgMap, prop); }

    // -----------------------------------------------------
    //                                      Foreign Property
    //                                      ----------------
    protected final Map<String, PropertyGateway> _efpgMap = newHashMap();
    { xsetupEfpg(); }
    @SuppressWarnings("unchecked")
    protected void xsetupEfpg() {
        setupEfpg(_efpgMap, et -> ((Product)et).getProductCategory(), (et, vl) -> ((Product)et).setProductCategory((OptionalEntity<ProductCategory>)vl), "productCategory");
        setupEfpg(_efpgMap, et -> ((Product)et).getProductStatus(), (et, vl) -> ((Product)et).setProductStatus((OptionalEntity<ProductStatus>)vl), "productStatus");
    }
    public PropertyGateway findForeignPropertyGateway(String prop)
    { return doFindEfpg(_efpgMap, prop); }

    // ===================================================================================
    //                                                                          Table Info
    //                                                                          ==========
    protected final String _tableDbName = "product";
    protected final String _tableDispName = "product";
    protected final String _tablePropertyName = "product";
    protected final TableSqlName _tableSqlName = new TableSqlName("product", _tableDbName);
    { _tableSqlName.xacceptFilter(DBFluteConfig.getInstance().getTableSqlNameFilter()); }
    public String getTableDbName() { return _tableDbName; }
    public String getTableDispName() { return _tableDispName; }
    public String getTablePropertyName() { return _tablePropertyName; }
    public TableSqlName getTableSqlName() { return _tableSqlName; }

    // ===================================================================================
    //                                                                         Column Info
    //                                                                         ===========
    protected final ColumnInfo _columnProductId = cci("product_id", "product_id", null, null, Integer.class, "productId", null, true, true, true, "serial", 10, 0, "nextval('product_product_id_seq'::regclass)", false, null, null, null, "purchaseList", null, false);
//"int64"
    protected final ColumnInfo _columnProductName = cci("product_name", "product_name", null, null, String.class, "productName", null, false, false, true, "varchar", 50, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnProductHandleCode = cci("product_handle_code", "product_handle_code", null, null, String.class, "productHandleCode", null, false, false, true, "varchar", 100, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnProductCategoryCode = cci("product_category_code", "product_category_code", null, null, String.class, "productCategoryCode", null, false, false, true, "bpchar", 3, 0, null, false, null, null, "productCategory", null, null, false);
//"string"
    protected final ColumnInfo _columnProductStatusCode = cci("product_status_code", "product_status_code", null, null, String.class, "productStatusCode", null, false, false, true, "bpchar", 3, 0, null, false, null, null, "productStatus", null, null, false);
//"string"
    protected final ColumnInfo _columnRegularPrice = cci("regular_price", "regular_price", null, null, Integer.class, "regularPrice", null, false, false, true, "int4", 10, 0, null, false, null, null, null, null, null, false);
//"int64"
    protected final ColumnInfo _columnRegisterDatetime = cci("register_datetime", "register_datetime", null, null, java.time.LocalDateTime.class, "registerDatetime", null, false, false, true, "timestamp", 26, 3, null, false, null, null, null, null, null, false);
//"df.Timestamp"
    protected final ColumnInfo _columnRegisterUser = cci("register_user", "register_user", null, null, String.class, "registerUser", null, false, false, true, "varchar", 200, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnRegisterProcess = cci("register_process", "register_process", null, null, String.class, "registerProcess", null, false, false, true, "varchar", 200, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnUpdateDatetime = cci("update_datetime", "update_datetime", null, null, java.time.LocalDateTime.class, "updateDatetime", null, false, false, true, "timestamp", 26, 3, null, false, null, null, null, null, null, false);
//"df.Timestamp"
    protected final ColumnInfo _columnUpdateUser = cci("update_user", "update_user", null, null, String.class, "updateUser", null, false, false, true, "varchar", 200, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnUpdateProcess = cci("update_process", "update_process", null, null, String.class, "updateProcess", null, false, false, true, "varchar", 200, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnVersionNo = cci("version_no", "version_no", null, null, Long.class, "versionNo", null, false, false, true, "int8", 19, 0, null, false, OptimisticLockType.VERSION_NO, null, null, null, null, false);
//"int64"

    /**
     * product_id: {PK, ID, NotNull, serial(10)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnProductId() { return _columnProductId; }
    /**
     * product_name: {IX, NotNull, varchar(50)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnProductName() { return _columnProductName; }
    /**
     * product_handle_code: {UQ, NotNull, varchar(100)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnProductHandleCode() { return _columnProductHandleCode; }
    /**
     * product_category_code: {NotNull, bpchar(3), FK to product_category}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnProductCategoryCode() { return _columnProductCategoryCode; }
    /**
     * product_status_code: {NotNull, bpchar(3), FK to product_status}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnProductStatusCode() { return _columnProductStatusCode; }
    /**
     * regular_price: {NotNull, int4(10)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRegularPrice() { return _columnRegularPrice; }
    /**
     * register_datetime: {NotNull, timestamp(26, 3)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRegisterDatetime() { return _columnRegisterDatetime; }
    /**
     * register_user: {NotNull, varchar(200)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRegisterUser() { return _columnRegisterUser; }
    /**
     * register_process: {NotNull, varchar(200)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRegisterProcess() { return _columnRegisterProcess; }
    /**
     * update_datetime: {NotNull, timestamp(26, 3)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnUpdateDatetime() { return _columnUpdateDatetime; }
    /**
     * update_user: {NotNull, varchar(200)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnUpdateUser() { return _columnUpdateUser; }
    /**
     * update_process: {NotNull, varchar(200)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnUpdateProcess() { return _columnUpdateProcess; }
    /**
     * version_no: {NotNull, int8(19)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnVersionNo() { return _columnVersionNo; }

    protected List<ColumnInfo> ccil() {
        List<ColumnInfo> ls = newArrayList();
        ls.add(columnProductId());
        ls.add(columnProductName());
        ls.add(columnProductHandleCode());
        ls.add(columnProductCategoryCode());
        ls.add(columnProductStatusCode());
        ls.add(columnRegularPrice());
        ls.add(columnRegisterDatetime());
        ls.add(columnRegisterUser());
        ls.add(columnRegisterProcess());
        ls.add(columnUpdateDatetime());
        ls.add(columnUpdateUser());
        ls.add(columnUpdateProcess());
        ls.add(columnVersionNo());
        return ls;
    }

    { initializeInformationResource(); }

    // ===================================================================================
    //                                                                         Unique Info
    //                                                                         ===========
    // -----------------------------------------------------
    //                                       Primary Element
    //                                       ---------------
    protected UniqueInfo cpui() { return hpcpui(columnProductId()); }
    public boolean hasPrimaryKey() { return true; }
    public boolean hasCompoundPrimaryKey() { return false; }

    // -----------------------------------------------------
    //                                        Unique Element
    //                                        --------------
    public UniqueInfo uniqueOf() { return hpcui(columnProductHandleCode()); }

    // ===================================================================================
    //                                                                       Relation Info
    //                                                                       =============
    // cannot cache because it uses related DB meta instance while booting
    // (instead, cached by super's collection)
    // -----------------------------------------------------
    //                                      Foreign Property
    //                                      ----------------
    /**
     * product_category by my product_category_code, named 'productCategory'.
     * @return The information object of foreign property. (NotNull)
     */
    public ForeignInfo foreignProductCategory() {
        Map<ColumnInfo, ColumnInfo> mp = newLinkedHashMap(columnProductCategoryCode(), ProductCategoryDbm.getInstance().columnProductCategoryCode());
        return cfi("fk_product_product_category", "productCategory", this, ProductCategoryDbm.getInstance(), mp, 0, org.dbflute.optional.OptionalEntity.class, false, false, false, false, null, null, false, "productList", false);
    }
    /**
     * product_status by my product_status_code, named 'productStatus'.
     * @return The information object of foreign property. (NotNull)
     */
    public ForeignInfo foreignProductStatus() {
        Map<ColumnInfo, ColumnInfo> mp = newLinkedHashMap(columnProductStatusCode(), ProductStatusDbm.getInstance().columnProductStatusCode());
        return cfi("fk_product_product_status", "productStatus", this, ProductStatusDbm.getInstance(), mp, 1, org.dbflute.optional.OptionalEntity.class, false, false, false, false, null, null, false, "productList", false);
    }

    // -----------------------------------------------------
    //                                     Referrer Property
    //                                     -----------------
    /**
     * purchase by product_id, named 'purchaseList'.
     * @return The information object of referrer property. (NotNull)
     */
    public ReferrerInfo referrerPurchaseList() {
        Map<ColumnInfo, ColumnInfo> mp = newLinkedHashMap(columnProductId(), PurchaseDbm.getInstance().columnProductId());
        return cri("fk_purchase_product", "purchaseList", this, PurchaseDbm.getInstance(), mp, false, "product");
    }

    // ===================================================================================
    //                                                                        Various Info
    //                                                                        ============
    public boolean hasSequence() { return true; }
    public String getSequenceName() { return "product_product_id_seq"; }
    public Integer getSequenceIncrementSize() { return 1; }
    public Integer getSequenceCacheSize() { return null; }
    public boolean hasVersionNo() { return true; }
    public ColumnInfo getVersionNoColumnInfo() { return _columnVersionNo; }

    // ===================================================================================
    //                                                                           Type Name
    //                                                                           =========
    public String getEntityTypeName() { return "com.mssoftech.dbflute.exentity.Product"; }
    public String getConditionBeanTypeName() { return "com.mssoftech.dbflute.cbean.ProductCB"; }
    public String getBehaviorTypeName() { return "com.mssoftech.dbflute.exbhv.ProductBhv"; }

    // ===================================================================================
    //                                                                         Object Type
    //                                                                         ===========
    public Class<Product> getEntityType() { return Product.class; }

    // ===================================================================================
    //                                                                     Object Instance
    //                                                                     ===============
    public Product newEntity() { return new Product(); }

    // ===================================================================================
    //                                                                   Map Communication
    //                                                                   =================
    public void acceptPrimaryKeyMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptPrimaryKeyMap((Product)et, mp); }
    public void acceptAllColumnMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptAllColumnMap((Product)et, mp); }
    public Map<String, Object> extractPrimaryKeyMap(Entity et) { return doExtractPrimaryKeyMap(et); }
    public Map<String, Object> extractAllColumnMap(Entity et) { return doExtractAllColumnMap(et); }
}

File: src/dbflute/adf/meta/productdbm.go
package meta

import (
	"github.com/mikeshimura/dbflute/df"
)

type ProductDbm_T struct {
	df.BaseDBMeta
	ColumnProductId *df.ColumnInfo
	ColumnProductName *df.ColumnInfo
	ColumnProductHandleCode *df.ColumnInfo
	ColumnProductCategoryCode *df.ColumnInfo
	ColumnProductStatusCode *df.ColumnInfo
	ColumnRegularPrice *df.ColumnInfo
	ColumnRegisterDatetime *df.ColumnInfo
	ColumnRegisterUser *df.ColumnInfo
	ColumnRegisterProcess *df.ColumnInfo
	ColumnUpdateDatetime *df.ColumnInfo
	ColumnUpdateUser *df.ColumnInfo
	ColumnUpdateProcess *df.ColumnInfo
	ColumnVersionNo *df.ColumnInfo
}

func (b *ProductDbm_T) GetProjectName() string {
	return df.DBCurrent_I.ProjectName
}

func (b *ProductDbm_T) GetDbCurrent() *df.DBCurrent {
	return df.DBCurrent_I
}

var ProductDbm *ProductDbm_T

func Create_ProductDbm() {
	ProductDbm = new(ProductDbm_T)
	ProductDbm.TableDbName = "product"
	ProductDbm.TableDispName = "product"
	ProductDbm.TablePropertyName = "product"
	ProductDbm.TableSqlName = new(df.TableSqlName)
	ProductDbm.TableSqlName.TableSqlName = "product"
	ProductDbm.TableSqlName.CorrespondingDbName = ProductDbm.TableDbName

	var product df.DBMeta
	product = ProductDbm
	ProductDbm.DBMeta=&product
	productIdSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo product_id
	productIdSqlName.ColumnSqlName = "product_id"
	productIdSqlName.IrregularChar = false
	ProductDbm.ColumnProductId = df.CCI(&product, "product_id", productIdSqlName, "", "", "Integer.class", "productId", "", true, true,true, "serial", 10, 0, "nextval('product_product_id_seq'::regclass)",false,"","", "","purchaseList","",false,"int64")
	productNameSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo product_name
	productNameSqlName.ColumnSqlName = "product_name"
	productNameSqlName.IrregularChar = false
	ProductDbm.ColumnProductName = df.CCI(&product, "product_name", productNameSqlName, "", "", "String.class", "productName", "", false, false,true, "varchar", 50, 0, "",false,"","", "","","",false,"string")
	productHandleCodeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo product_handle_code
	productHandleCodeSqlName.ColumnSqlName = "product_handle_code"
	productHandleCodeSqlName.IrregularChar = false
	ProductDbm.ColumnProductHandleCode = df.CCI(&product, "product_handle_code", productHandleCodeSqlName, "", "", "String.class", "productHandleCode", "", false, false,true, "varchar", 100, 0, "",false,"","", "","","",false,"string")
	productCategoryCodeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo product_category_code
	productCategoryCodeSqlName.ColumnSqlName = "product_category_code"
	productCategoryCodeSqlName.IrregularChar = false
	ProductDbm.ColumnProductCategoryCode = df.CCI(&product, "product_category_code", productCategoryCodeSqlName, "", "", "String.class", "productCategoryCode", "", false, false,true, "bpchar", 3, 0, "",false,"","", "productCategory","","",false,"string")
	productStatusCodeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo product_status_code
	productStatusCodeSqlName.ColumnSqlName = "product_status_code"
	productStatusCodeSqlName.IrregularChar = false
	ProductDbm.ColumnProductStatusCode = df.CCI(&product, "product_status_code", productStatusCodeSqlName, "", "", "String.class", "productStatusCode", "", false, false,true, "bpchar", 3, 0, "",false,"","", "productStatus","","",false,"string")
	regularPriceSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo regular_price
	regularPriceSqlName.ColumnSqlName = "regular_price"
	regularPriceSqlName.IrregularChar = false
	ProductDbm.ColumnRegularPrice = df.CCI(&product, "regular_price", regularPriceSqlName, "", "", "Integer.class", "regularPrice", "", false, false,true, "int4", 10, 0, "",false,"","", "","","",false,"int64")
	registerDatetimeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo register_datetime
	registerDatetimeSqlName.ColumnSqlName = "register_datetime"
	registerDatetimeSqlName.IrregularChar = false
	ProductDbm.ColumnRegisterDatetime = df.CCI(&product, "register_datetime", registerDatetimeSqlName, "", "", "java.time.LocalDateTime.class", "registerDatetime", "", false, false,true, "timestamp", 26, 3, "",false,"","", "","","",false,"df.Timestamp")
	registerUserSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo register_user
	registerUserSqlName.ColumnSqlName = "register_user"
	registerUserSqlName.IrregularChar = false
	ProductDbm.ColumnRegisterUser = df.CCI(&product, "register_user", registerUserSqlName, "", "", "String.class", "registerUser", "", false, false,true, "varchar", 200, 0, "",false,"","", "","","",false,"string")
	registerProcessSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo register_process
	registerProcessSqlName.ColumnSqlName = "register_process"
	registerProcessSqlName.IrregularChar = false
	ProductDbm.ColumnRegisterProcess = df.CCI(&product, "register_process", registerProcessSqlName, "", "", "String.class", "registerProcess", "", false, false,true, "varchar", 200, 0, "",false,"","", "","","",false,"string")
	updateDatetimeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo update_datetime
	updateDatetimeSqlName.ColumnSqlName = "update_datetime"
	updateDatetimeSqlName.IrregularChar = false
	ProductDbm.ColumnUpdateDatetime = df.CCI(&product, "update_datetime", updateDatetimeSqlName, "", "", "java.time.LocalDateTime.class", "updateDatetime", "", false, false,true, "timestamp", 26, 3, "",false,"","", "","","",false,"df.Timestamp")
	updateUserSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo update_user
	updateUserSqlName.ColumnSqlName = "update_user"
	updateUserSqlName.IrregularChar = false
	ProductDbm.ColumnUpdateUser = df.CCI(&product, "update_user", updateUserSqlName, "", "", "String.class", "updateUser", "", false, false,true, "varchar", 200, 0, "",false,"","", "","","",false,"string")
	updateProcessSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo update_process
	updateProcessSqlName.ColumnSqlName = "update_process"
	updateProcessSqlName.IrregularChar = false
	ProductDbm.ColumnUpdateProcess = df.CCI(&product, "update_process", updateProcessSqlName, "", "", "String.class", "updateProcess", "", false, false,true, "varchar", 200, 0, "",false,"","", "","","",false,"string")
	versionNoSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo version_no
	versionNoSqlName.ColumnSqlName = "version_no"
	versionNoSqlName.IrregularChar = false
	ProductDbm.ColumnVersionNo = df.CCI(&product, "version_no", versionNoSqlName, "", "", "Long.class", "versionNo", "", false, false,true, "int8", 19, 0, "",false,"OptimisticLockType.VERSION_NO","", "","","",false,"int64")

	ProductDbm.ColumnInfoList = new(df.List)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnProductId)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnProductName)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnProductHandleCode)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnProductCategoryCode)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnProductStatusCode)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnRegularPrice)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnRegisterDatetime)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnRegisterUser)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnRegisterProcess)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnUpdateDatetime)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnUpdateUser)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnUpdateProcess)
	ProductDbm.ColumnInfoList.Add(ProductDbm.ColumnVersionNo)


	ProductDbm.ColumnInfoMap=make(map[string]int)
	ProductDbm.ColumnInfoMap["productId"]=0
		ProductDbm.ColumnInfoMap["productName"]=1
		ProductDbm.ColumnInfoMap["productHandleCode"]=2
		ProductDbm.ColumnInfoMap["productCategoryCode"]=3
		ProductDbm.ColumnInfoMap["productStatusCode"]=4
		ProductDbm.ColumnInfoMap["regularPrice"]=5
		ProductDbm.ColumnInfoMap["registerDatetime"]=6
		ProductDbm.ColumnInfoMap["registerUser"]=7
		ProductDbm.ColumnInfoMap["registerProcess"]=8
		ProductDbm.ColumnInfoMap["updateDatetime"]=9
		ProductDbm.ColumnInfoMap["updateUser"]=10
		ProductDbm.ColumnInfoMap["updateProcess"]=11
		ProductDbm.ColumnInfoMap["versionNo"]=12
	    ProductDbm.PrimaryKey = true
    ProductDbm.CompoundPrimaryKey = false
	ui := new(df.UniqueInfo)
	ui.DbMeta = &product
	ui.Primary = true
	ui.UniqueColumnList = new(df.List)
	ui.UniqueColumnList.Add(ProductDbm.ColumnProductId)

	ProductDbm.PrimaryInfo = new(df.PrimaryInfo)
	ProductDbm.PrimaryInfo.UniqueInfo = ui
	
	ProductDbm.SequenceFlag = true
	ProductDbm.SequenceName = "product_product_id_seq"
	ProductDbm.SequenceIncrementSize = 1
	ProductDbm.SequenceCacheSize = 0
	ProductDbm.VersionNoFlag = true
	ProductDbm.VersionNoColumnInfo = ProductDbm.ColumnVersionNo
	
	var productMeta df.DBMeta = ProductDbm
	df.DBMetaInstanceHandler_I.TableDbNameInstanceMap["Product"] = &productMeta
}
